adding fp16 tests in test_nn

ailzhang opened the pull request:
The first part to fix issue #4940 . planning to update test_cuda.py next. 
Uploading a draft version based on discussion with @gchanan for further discussing the best way to scan different data types for the same test, and some potential bugs in fp16(need to confirm)

One drawback about using decorators is that for different date types on the same test case, setUp() function is only run once. An alternative way it to add test for every single data type as a test case separately as we did in test_cuda.py.

Commit:
first draft of adding fp16 tests in test_nn

apaszke commented on test/test_nn.py:
```diff
@@ -35,6 +35,28 @@
+def repeat_test_for_types(dtypes):
+    def repeatHelper(f):
+        def callHelper(*args):
+            for dtype in dtypes:
```
One suggestion would be to use `subTest` so that we see for which type the failure occured (but you need to guard it so that it's only used in Python 3.4+

Commit:
Use subTest;Fix TODOs

ailzhang commented:
* Added subTest
* Fixed TODOs. 
* Trying to implement half tensor testcases with minimal code change. This file needs to be refactored in the future. 
This PR is ready for review, thanks! :) @apaszke @ezyang

Commit:
add python version guard for subTest

csarofeen commented:
@ailzhang this PR looks great, thanks for working on this!

apaszke commented on test/test_nn.py:
```diff
@@ -285,6 +298,20 @@ def _get_target(self):
 
 class TestNN(NNTestCase):
+    # Used for run the same test with different tensor types
+    def repeat_test_for_types(dtypes):
```
If you want to put it as a regular function, but in a class, then you need to decorate it with `@staticfunction`

apaszke commented on test/test_nn.py:
```diff
@@ -285,6 +298,20 @@ def _get_target(self):
+                        f(self, *args, dtype=dtype)
+
+            return callHelper
+        return repeatHelper
```
nit: we normally use `underscore_case` for variable and function names

apaszke commented:
This looks good, but do we plan to add support for half tests to the auto-generated tests as well?

Commit:
address comments

apaszke commented on test/test_nn.py:
```diff
@@ -1202,16 +1232,17 @@ def test_gumbel_softmax_st(self):
         self._test_gumbel_softmax_st(False)
 
     @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    # torch.nn.functional doesn't provide half() method
```
I'm not sure if I understand the comment. It doesn't and it won't ever provide such a method. It should work with all types just fine

apaszke reviewed (commented):
Looks good. Two minor things

Commit:
revert missing lines;fix test_gumbel_softmax_st_cuda

Commit:
resolve conflicts

Commit:
fix test_embedding_bag

Commit:
resolve conflict with master branch

ailzhang commented:
@apaszke  🙏

apaszke commented on test/test_nn.py:
```diff
@@ -2451,7 +2507,7 @@ def test_Conv2d_inconsistent_types_on_GPU_without_cudnn(self):
         inputs = Variable(torch.randn(4, 1, 7, 7).float().cuda())
         weights = Variable(torch.randn(1, 1, 3, 3).double().cuda())
-        bias = Variable(torch.randn(1).double().cuda())
+        bias = Variable(torch.randn(1).half().cuda())
```
Let's leave this one in `double()` (same for the test above)

apaszke reviewed (approved):
Can you only revert this `double()` -> `half()` change in two tests? Looks good

Commit:
revert half to double in test_Conv2d_inconsistent_types_on_GPU*
